// genesys-provisioning-client-js/src/internal/code-gen/provisioning-api/model/InboundData.js
/**
 * Provisioning API
 * The provisioning service will be used to create objects in configuration server in a way that is consistent with the BEC model. The \"users\" resource will be used to perform CRUD operations on config server person objects. Note that this API will extend the low level configuration server API to add additional functionality that will allow persons to be created in a way that makes them immediately useable in BEC. Operations on this resource may result in other configuration objects being modified as well. 
 *
 * OpenAPI spec version: 9.0.000.26.2572
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 *
 * Swagger Codegen version: 2.3.1
 *
 * Do not edit the class manually.
 *
 */

(function(root, factory) {
  if (typeof define === 'function' && define.amd) {
    // AMD. Register as an anonymous module.
    define(['ApiClient'], factory);
  } else if (typeof module === 'object' && module.exports) {
    // CommonJS-like environments that support module.exports, like Node.
    module.exports = factory(require('../ApiClient'));
  } else {
    // Browser globals (root is window)
    if (!root.ProvisioningApi) {
      root.ProvisioningApi = {};
    }
    root.ProvisioningApi.InboundData = factory(root.ProvisioningApi.ApiClient);
  }
}(this, function(ApiClient) {
  'use strict';




  /**
   * The InboundData model module.
   * @module model/InboundData
   * @version 9.0.000.26.2572
   */

  /**
   * Constructs a new <code>InboundData</code>.
   * @alias module:model/InboundData
   * @class
   * @param name {String} Name of an inbound client 
   */
  var exports = function(name) {
    var _this = this;

    _this['name'] = name;







  };

  /**
   * Constructs a <code>InboundData</code> from a plain JavaScript object, optionally creating a new instance.
   * Copies all relevant properties from <code>data</code> to <code>obj</code> if supplied or a new instance if not.
   * @param {Object} data The plain JavaScript object bearing properties of interest.
   * @param {module:model/InboundData} obj Optional instance to populate.
   * @return {module:model/InboundData} The populated <code>InboundData</code> instance.
   */
  exports.constructFromObject = function(data, obj) {
    if (data) {
      obj = obj || new exports();

      if (data.hasOwnProperty('name')) {
        obj['name'] = ApiClient.convertToType(data['name'], 'String');
      }
      if (data.hasOwnProperty('maxMessageSize')) {
        obj['maxMessageSize'] = ApiClient.convertToType(data['maxMessageSize'], 'Number');
      }
      if (data.hasOwnProperty('mailbox')) {
        obj['mailbox'] = ApiClient.convertToType(data['mailbox'], 'String');
      }
      if (data.hasOwnProperty('password')) {
        obj['password'] = ApiClient.convertToType(data['password'], 'String');
      }
      if (data.hasOwnProperty('host')) {
        obj['host'] = ApiClient.convertToType(data['host'], 'String');
      }
      if (data.hasOwnProperty('port')) {
        obj['port'] = ApiClient.convertToType(data['port'], 'Number');
      }
      if (data.hasOwnProperty('connectionSecurity')) {
        obj['connectionSecurity'] = ApiClient.convertToType(data['connectionSecurity'], 'String');
      }
      if (data.hasOwnProperty('type')) {
        obj['type'] = ApiClient.convertToType(data['type'], 'String');
      }
    }
    return obj;
  }

  /**
   * Name of an inbound client 
   * @member {String} name
   */
  exports.prototype['name'] = undefined;
  /**
   * @member {Number} maxMessageSize
   */
  exports.prototype['maxMessageSize'] = undefined;
  /**
   * @member {String} mailbox
   */
  exports.prototype['mailbox'] = undefined;
  /**
   * @member {String} password
   */
  exports.prototype['password'] = undefined;
  /**
   * @member {String} host
   */
  exports.prototype['host'] = undefined;
  /**
   * @member {Number} port
   */
  exports.prototype['port'] = undefined;
  /**
   * @member {String} connectionSecurity
   */
  exports.prototype['connectionSecurity'] = undefined;
  /**
   * POP3 or IMAP 
   * @member {String} type
   */
  exports.prototype['type'] = undefined;



  return exports;
}));
